fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/typing/TypeRepository.java

```diff
@@ -471,19 +471,46 @@ public Builder setPackage(@Nonnull final String name) {
         @Nonnull
         public Builder addTypeIfNeeded(@Nonnull final Type type) {
             if (!typeToNameMap.containsKey(type)) {
+                // Check if we have a structurally identical type with different nullability already registered
+                Type canonicalType = findCanonicalTypeForStructure(type);
+                if (canonicalType != null) {
+                    // Use the same protobuf name as the canonical type
+                    String existingProtoName = typeToNameMap.get(canonicalType);
+                    if (existingProtoName != null) {
+                        typeToNameMap.put(type, existingProtoName);
+                        return this;
+                    }
+                }
+
+                // Standard case: define the protobuf type
                 type.defineProtoType(this);
             }
             return this;
         }
 
+        /**
+         * Finds a type in typeToNameMap that has the same structure as the given type but different nullability.
+         * Returns null if no such type exists.
+         */
+        @Nullable
+        private Type findCanonicalTypeForStructure(@Nonnull final Type type) {
+            for (Map.Entry<Type, String> entry : typeToNameMap.entrySet()) {
+                Type existingType = entry.getKey();
+                if (differsOnlyInNullability(type, existingType)) {
+                    return existingType;
+                }
+            }
+            return null;
+        }
+
         @Nonnull
         public Optional<String> getTypeName(@Nonnull final Type type) {
             return Optional.ofNullable(typeToNameMap.get(type));
         }
 
         @Nonnull
         public Optional<Type> getTypeByName(@Nonnull final String name) {
-            return Optional.ofNullable(typeToNameMap.inverse().get(name));
+            return Optional.ofNullable(nameToCanonicalTypeMap.get(name));
```

Comment:
I think this method is unused within the code base. Should we just get rid of it?

Reply:
Sure 👍